File: src/main/java/com/github/oxyethylene/otelcollector/grpc/MetricsServiceImpl.java
package com.github.oxyethylene.otelcollector.grpc;

import io.grpc.stub.StreamObserver;
import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceRequest;
import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceResponse;
import io.opentelemetry.proto.collector.metrics.v1.MetricsServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class MetricsServiceImpl extends MetricsServiceGrpc.MetricsServiceImplBase {
    private static final Logger log = LoggerFactory.getLogger(MetricsServiceImpl.class);

    @Override
    public void export(ExportMetricsServiceRequest request,
                       StreamObserver<ExportMetricsServiceResponse> responseObserver) {
        log.info("Received OTLP metrics export");
        ExportMetricsServiceResponse resp = ExportMetricsServiceResponse.getDefaultInstance();
        responseObserver.onNext(resp);
        responseObserver.onCompleted();
    }
}

File: src/main/kotlin/com/github/oxyethylene/otelcollector/grpc/MetricsServiceImpl.kt
package com.github.oxyethylene.otelcollector.grpc

import com.github.oxyethylene.otelcollector.model.StoredMetricPoint
import com.github.oxyethylene.otelcollector.storage.MetricStore
import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceRequest
import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceResponse
import io.opentelemetry.proto.collector.metrics.v1.MetricsServiceGrpcKt
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class MetricsServiceImpl(private val metricStore: MetricStore) : MetricsServiceGrpcKt.MetricsServiceCoroutineImplBase() {
    override suspend fun export(request: ExportMetricsServiceRequest): ExportMetricsServiceResponse {
        log.debug("Received OTLP metrics export")
        for (resourceMetrics in request.resourceMetricsList) {
            val resource = resourceMetrics.resource
            for (scopeMetrics in resourceMetrics.scopeMetricsList) {
                val scopeName = scopeMetrics.scope.name
                for (metric in scopeMetrics.metricsList) {
                    StoredMetricPoint.fromProto(metric, resource, scopeName)
                        .forEach(metricStore::store)
                }
            }
        }
        return ExportMetricsServiceResponse.getDefaultInstance()
    }

    companion object {
        private val log = LoggerFactory.getLogger(MetricsServiceImpl::class.java)
    }
}
